### ISSUE TITLE
Freeze on netlink error reply?

### ISSUE THREAD

**Issue by grinapo** [Tomas-M/iotop#28]
iotop enters an infinite wait on netlink, probably as a result of ignoring a previous error:

```
11:31:14.201289 socket(AF_NETLINK, SOCK_RAW, NETLINK_GENERIC) = 3
11:31:14.201393 bind(3, {sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, 12) = 0
11:31:14.201511 getpid()                = 46880
11:31:14.201610 sendto(3, {{len=36, type=nlctrl, flags=NLM_F_REQUEST, seq=0, pid=46880}, "\x03\x01\x00\x00\x0e\x00\x02\x00\x54\x41\x53\x4b\x53\x54\x41\x54\x53\x00\x00\x00"}, 36, 0, {sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, 12) = 36
11:31:14.202179 recvfrom(3, {{len=56, type=NLMSG_ERROR, flags=0, seq=0, pid=46880}, {error=-ENOENT, msg={{len=36, type=nlctrl, flags=NLM_F_REQUEST, seq=0, pid=46880}, "\x03\x01\x00\x00\x0e\x00\x02\x00\x54\x41\x53\x4b\x53\x54\x41\x54\x53\x00\x00\x00"}}}, 1044, 0, NULL, NULL) = 56
11:31:14.202340 rt_sigaction(SIGINT, {sa_handler=0x565546aa1060, sa_mask=[INT], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7fc8a8c1cd60}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
11:31:14.202472 openat(AT_FDCWD, "/proc", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 4
11:31:14.202597 fstat(4, {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
11:31:14.202738 getdents64(4, 0x56554726c380 /* 90 entries */, 32768) = 2632
11:31:14.202978 openat(AT_FDCWD, "/proc/1/task", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 5
11:31:14.203094 fstat(5, {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
11:31:14.203202 getdents64(5, 0x5655472743c0 /* 3 entries */, 32768) = 72
11:31:14.203309 sendto(3, {{len=28, type=0 /* NLMSG_??? */, flags=NLM_F_REQUEST, seq=0, pid=1}, "\x01\x01\x00\x00\x08\x00\x01\x00\x01\x00\x00\x00"}, 28, 0, {sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, 12) = 28
11:31:14.203458 recvfrom(3,   …
```

This is an lxc container and it may restrict access to various things around. Unhandled errors may result in weird behaviours.

**Comment by bbonev** [comment]
Thanks for the report!

Can you help a little bit with how to reproduce that? I am not using lxc containers and am not well aware with them - is it an unprivileged or a privileged one? Are there some specific settings?

**Comment by grinapo** [comment]
I am sure it will do the same on any unprivileged lxc container, but I cannot think any more specific way to reproduce.
If you're fatally stuck try to contact me privately (like `@grin:grin.hu` on matrix) and I'll create a container for you to play in.

**Comment by bbonev** [comment]
**TL;DR - iotop can not run in unprivileged namespaces**

I didn't even try to run lxc containers - too much hassle... It is the same in docker.

The problem with unprivileged namespaces is that netlink is not available there at all - unlike the normal network stack and the related netlink, the part of netlink that is used by iotop is not virtualized in current kernels. If some shiny day kernel support for that gets implemented then iotop will automatically start working in such namespaces.

From this `sendto(3, {{len=28, type=0 /* NLMSG_??? */,` it is seen that the numeric family id for TASKSTATS_GENL_NAME ("TASKSTATS") could not be fetched, `nl_fam_id` is 0 and the `sendto` call is generating an invalid message to which the kernel will not reply, hence the freeze.

I have improved the error handling, as a result iotop will exit with error when run in such namespaces. You can find the result here: https://github.com/bbonev/iotop

Most probably I will improve the error message before releasing a new version...
